src/wallet/wallet.cpp

```diff
@@ -857,80 +858,128 @@ bool CWallet::LoadUnifiedAddressMetadata(const ZcashdUnifiedAddressMetadata &add
+            } else {
+                return std::make_pair(PaymentAddress{addr}, RecipientType::WalletExternalAddress);
+            }
+        },
+        [&](const CScriptID& addr) {
+            auto ua = self->FindUnifiedAddressByReceiver(addr);
+            if (ua.has_value()) {
+                return std::make_pair(PaymentAddress{ua.value()}, RecipientType::WalletExternalAddress);
+            } else if (HaveCScript(addr)) {
+                // we never generate internal p2sh addresses, so all are wallet external
+                return std::make_pair(PaymentAddress{addr}, RecipientType::WalletExternalAddress);
+            }
+
+            return std::make_pair(PaymentAddress{addr}, RecipientType::CounterpartyAddress);
+        },
+        [&](const SaplingPaymentAddress& addr) {
+            auto ua = self->FindUnifiedAddressByReceiver(addr);
+            if (ua.has_value()) {
+                // UAs are always external addresses
+                return std::make_pair(PaymentAddress{ua.value()}, RecipientType::WalletExternalAddress);
+            } else if (ufvk.has_value() && ufvk->GetSaplingKey().has_value()) {
+                auto saplingKey = ufvk->GetSaplingKey().value();
+                auto j = saplingKey.DecryptDiversifier(addr);
+                if (j.has_value()) {
+                    // external addresses correspond to UAs.
+                    if (j.value().second) {
+                        // std::get is safe here because we know we have a valid Sapling diversifier index
+                        auto defaultUA = std::get<std::pair<UnifiedAddress, diversifier_index_t>>(
+                                ufvk->Address(j.value().first, CWallet::DefaultReceiverTypes(nHeight)));
+                        return std::make_pair(PaymentAddress{defaultUA.first}, RecipientType::WalletExternalAddress);
+                    } else {
+                        return std::make_pair(PaymentAddress{addr}, RecipientType::WalletInternalAddress);
+                    }
                 }
-            },
-            [&](const OrchardRawAddress& addr) {
-                auto ua = self->FindUnifiedAddressByReceiver(addr);
-                if (ua.has_value()) {
-                    return libzcash::PaymentAddress{ua.value()};
-                } else if (ufvk.has_value() && ufvk->GetOrchardKey().has_value()) {
-                    auto orchardKey = ufvk->GetOrchardKey().value();
-                    auto j = orchardKey.ToIncomingViewingKey().DecryptDiversifier(addr);
-                    if (j.has_value()) {
-                        auto genResult = ufvk->Address(j.value(), CWallet::DefaultReceiverTypes());
+            }
+
+            // We don't produce internal change addresses for legacy Sapling
+            // addresses, so this must be a counterparty address
+            return std::make_pair(PaymentAddress{addr}, RecipientType::CounterpartyAddress);
```

Comment:
This does not hold, because AFAICT we have not checked whether the address is a legacy Sapling address we generated via `z_getnewaddress` (in which case it would be `RecipientType::WalletExternalAddress`).

Reply:
Fixed in https://github.com/zcash/zcash/pull/5715/commits/151074cbb13455afb8257354c0d856a79da3f332